sensor/common/processsignal/pipeline.go

```diff
@@ -72,20 +110,28 @@ func (p *Pipeline) Process(signal *storage.ProcessSignal) {
 }
 
 func (p *Pipeline) sendIndicatorEvent() {
-	for indicator := range p.enrichedIndicators {
+	for indicator := range p.cm.GetOutput() {
 		if !p.processFilter.Add(indicator) {
 			continue
 		}
 		p.detector.ProcessIndicator(indicator)
-
-		p.indicators <- message.New(&central.MsgFromSensor{Msg: &central.MsgFromSensor_Event{Event: &central.SensorEvent{
-			Id:     indicator.GetId(),
-			Action: central.ResourceAction_CREATE_RESOURCE,
-			Resource: &central.SensorEvent_ProcessIndicator{
-				ProcessIndicator: indicator,
+		msg := message.New(&central.MsgFromSensor{
+			Msg: &central.MsgFromSensor_Event{
+				Event: &central.SensorEvent{
+					Id:     indicator.GetId(),
+					Action: central.ResourceAction_CREATE_RESOURCE,
+					Resource: &central.SensorEvent_ProcessIndicator{
+						ProcessIndicator: indicator,
+					},
+				},
 			},
-		},
-		},
 		})
+		if !p.centralReady.IsDone() {
+			log.Debugf("Handling ProcessSignal in Offline mode: %s", msg.String())
+			canceledCtx, cancel := context.WithCancel(msg.Context)
+			cancel()
+			msg.Context = canceledCtx
```

Comment:
Additionally, at this point, if you know that the message should not be processed, why not avoid sending to the channel rather than appending a purposefully canceled context?

IMO the code should be something like:
```go
p.indicators <- message.NewExpiring(msg, p.currentContext) // always send message with current context
```

And then the Notify:
```go
func (p *Pipeline) Notify(e common.SensorComponentEvent) {
	switch e {
	case common.SensorComponentEventCentralReachable:
                // ... 
                p.createNewContext()
	case common.SensorComponentEventOfflineMode:
                // ...
                p.cancelCurrentContext() // expire any messages sent with this context
```

Reply:
+1 to this

Reply:
Sorry for the confusion. 

Let me try to clarify my thoughts. I see two potential improvements:

1) The message might expire **after** you sent it. For this reason the suggestion is to keep a context property in the component and always pass it to the `ExpiringMessage` constructor. This way, once the message is canceled, it would be picked up at _any point_ in the pipeline (this is related to the code suggestion above). 

2) As an improvement, since you've added an `if` here to check if the component is in offline mode before sending, I suggest to -- rather than explicitly setting a canceled context -- to just skip sending into the queue. (this is not included in the suggestion snippet).

A combination of both would look like:

```go
if !p.centralReady.IsDone() {
        p.indicators <- message.NewExpiring(msg, p.currentContext) // always send message with current context
}
```

(2) is a slight performance improvement, since you will spare the following component to having to work on the message and (maybe) check the context. But it is not needed, and I don't think other components are doing that. I'll leave it up to you.

> that we will not change the routine but just assign an already canceled context.

What I meant by this is suggestion (1). If the component always assign it's current context to the message and during it's routine the context gets canceled, we will assign a canceled context to the message.

Reply:
I implemented suggestion 1) for now. Adding 2) is easy but it would complicate the tests a bit.
Let me know if I should go for it despite that.